cpp/src/rolling/rolling_detail.cuh

```diff
@@ -64,386 +66,703 @@
 namespace cudf {
 
 namespace detail {
+
 namespace {  // anonymous
+
 /**
- * @brief Only COUNT_VALID operation is executed and count is updated
- *        depending on `min_periods` and returns true if it was
- *        valid, else false.
+ * @brief Operator for applying a generic (non-specialized) rolling aggregation on a single window.
  */
-template <typename InputType,
-          typename OutputType,
-          typename agg_op,
-          aggregation::Kind op,
-          bool has_nulls,
-          std::enable_if_t<op == aggregation::COUNT_VALID>* = nullptr>
-bool __device__ process_rolling_window(column_device_view input,
-                                       column_device_view ignored_default_outputs,
-                                       mutable_column_device_view output,
-                                       size_type start_index,
-                                       size_type end_index,
-                                       size_type current_index,
-                                       size_type min_periods)
-{
-  // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
-  // for CUDA 10.0 and below (fixed in CUDA 10.1)
-  volatile cudf::size_type count = 0;
+template <typename InputType, aggregation::Kind op>
+struct DeviceRolling {
+  size_type min_periods;
 
-  bool output_is_valid = ((end_index - start_index) >= min_periods);
+  // what operations do we support
+  template <typename T = InputType, aggregation::Kind O = op>
+  static constexpr bool is_supported()
+  {
+    return cudf::detail::is_valid_aggregation<T, O>() && has_corresponding_operator<O>() &&
+             // MIN/MAX supports all fixed width types
+             ((O == aggregation::MIN || O == aggregation::MAX) && cudf::is_fixed_width<T>()) ||
 
-  if (output_is_valid) {
-    if (!has_nulls) {
-      count = end_index - start_index;
-    } else {
-      count = thrust::count_if(thrust::seq,
-                               thrust::make_counting_iterator(start_index),
-                               thrust::make_counting_iterator(end_index),
-                               [&input](auto i) { return input.is_valid_nocheck(i); });
+           // SUM supports all fixed width types except timestamps
+           ((O == aggregation::SUM) && (cudf::is_fixed_width<T>() && !cudf::is_timestamp<T>())) ||
+
+           // MEAN supports numeric and duration
+           ((O == aggregation::MEAN) && (cudf::is_numeric<T>() || cudf::is_duration<T>()));
+  }
```

Comment:
I thought the goal was to make it so each operation owns the information about what types it supports. This code seems contradictory to that goal.

Reply:
Meaning, split this into DeviceRollingMean, DeviceRollingMax, DeviceRollingSum, etc?  I suppose we could do that but the meat of this is the code in the operator(), so we'd either have to duplicate that or make DeviceRolling a base class and derive these.  I opted for this because it's less code to sort through.  But I can change if you'd like.

Reply:
Resolved offline.  Conclusion:  leave extra logic for now and remove after some additional changes to `is_valid_aggregation`